---
title: "16S Analysis"
author: "KH Wong, ES Strand, and AH Huffmyer"
output:
  github_document: default
  pdf_document:
    keep_tex: yes
  html_document:
    toc: yes
    toc_depth: 6
    toc_float: yes
editor_options: 
  chunk_output_type: console
---

Script to analyze 16S sequences for the Porites July Bleaching project. 

Following this tutorial: https://www.yanh.org/2021/01/01/microbiome-r/

# Setup

Set up workspace, set options, and load required packages.    
```{r}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```


# Installing Packages

```{r}
p1 <- c("tidyverse", "vegan", "BiocManager")
p2 <- c("phyloseq", "ANCOMBC", "DESeq2", "ComplexHeatmap")
load_package <- function(p) {
  if (!requireNamespace(p, quietly = TRUE)) {
    ifelse(p %in% p1, 
           install.packages(p, repos = "http://cran.us.r-project.org/"), 
           BiocManager::install(p))
  }
  library(p, character.only = TRUE, quietly = TRUE)
}
invisible(lapply(c(p1,p2), load_package))

library(lme4)
library(lmerTest)
library(ggpubr)
library(broom)
library(DESeq2)

```

# Build phyloseq project

## Importing dataframes, cleaning taxonomy files, and filtering out Chloroplasts and non-bacterial assemblages

```{r, echo=FALSE, warning = FALSE}

otu <- read.table(file = "output/16S/processed_data/feature-table.tsv", sep = "\t", header = T, row.names = 1, skip = 1, comment.char = "")

taxonomy <- read.table(file = "output/16S/processed_data/taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

# clean the taxonomy, Greengenes format
tax <- taxonomy %>%
  dplyr::select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}

tax.clean2 <- tax.clean %>% 
  filter(Kingdom == "Bacteria") %>% 
  filter(Order != "Chloroplast")
```


## Creating Phyloseq objects

```{r}
metadata <- read.table(file = "output/16S/metadata.tsv", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)
TAX = tax_table(as.matrix(tax.clean2))
SAMPLE <- sample_data(metadata)
SAMPLE$Timepoint <- as.factor(SAMPLE$Timepoint)
SAMPLE$Group <- as.factor(SAMPLE$Group)
names(SAMPLE)[3] <- "Day"
SAMPLE$GroupDay <- paste(SAMPLE$Group, SAMPLE$Day, sep = "-")
TREE = read_tree("output/16S/processed_data/tree.nwk")
# merge the data
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)
ps

```

# Check the sequencing depth with rarefaction curves

From the QIIME2 output, there are no samples that drop off with sampling depth. 2000 and is the sample size that keeps all but 1 of my samples after testing

```{r, echo=FALSE, warning = FALSE}
set.seed(111) # keep result reproductive
ps.rarefied = rarefy_even_depth(ps, rngseed=1, sample.size=2300, replace=F)
ps.rarefied
```

`set.seed(1)` was used to initialize repeatable random subsampling.
Please record this for your records so others can reproduce.
Try `set.seed(1); .Random.seed` for the full vector
...
1 samples removedbecause they contained fewer reads than `sample.size`.
Up to first five removed samples are: 

WSH154
...
1609OTUs were removed because they are no longer 
present in any sample after random subsampling


# Creating a dataframe for omic integration

Renaming ASV ID and sample names. There is probably a more elegant way to do this...

```{r, echo=FALSE, warning = FALSE}
#Assigning ASV numbers to ASV ID
ps.rarefied_names <- as.data.frame(taxa_names(ps.rarefied))
names(ps.rarefied_names)[1] <- "ASV_ID"
ps.rarefied_names$Rownum <- as.numeric(rownames(ps.rarefied_names))
ps.rarefied_names$ASV_ASV <- "ASV"
ps.rarefied_names$ASV <- paste(ps.rarefied_names$ASV_ASV, ps.rarefied_names$Rownum, sep = "-")
ps.rarefied_names2 <- ps.rarefied_names %>% dplyr::select(ASV_ID, ASV)

# Extract abundance matrix from the phyloseq object
OTU1 = as(otu_table(ps.rarefied), "matrix")
# transpose if necessary
#if(taxa_are_rows(ps.rarefied)){OTU1 <- t(OTU1)}
# Coerce to data.frame
OTUdf = as.data.frame(OTU1)
#Make rownames into a column
OTUdf2 <- tibble::rownames_to_column(OTUdf, "ASV_ID")

# Merging dataframes to get ASV
OTUdf3 <- merge(OTUdf2, ps.rarefied_names2, by = "ASV_ID") %>% dplyr::select(-ASV_ID)
rownames(OTUdf3) <- paste0(OTUdf3$ASV)
OTUdf4 <- OTUdf3 %>% dplyr::select(-ASV)

#Formatting the Sample Datasheet to rename the samples
metadata2 <- tibble::rownames_to_column(metadata, "Sample") 
metadata2$ColDay <- paste(metadata2$Colony_ID, metadata2$Timepoint, sep = "-")
metadata3 <- metadata2 %>% dplyr::select(Sample, ColDay)
# 
# #Reformatting OTU dataframe and merge with sample info
# OTUdf5 <- as.data.frame(t(OTUdf4))
# OTUdf6 <- tibble::rownames_to_column(OTUdf5, "Sample")
# OTUdf7 <- merge(OTUdf6, metadata3, by = "Sample") %>% dplyr::select(-Sample)
# rownames(OTUdf7) <- paste0(OTUdf7$ColDay)
# OTUdf_final <- OTUdf7 %>% dplyr::select(-ColDay)

###Create a DESeqDataSet design from OTU count matrix and labels. Here we set the design to look at Day and time.

metadata2 <- metadata2 %>% filter(Sample != "WSH154")

#Checking that all row and column names match. Should return "TRUE"
all(rownames(metadata2$Sample) %in% colnames(OTUdf4))
all(rownames(metadata2$Sample) == colnames(OTUdf4)) 

#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = round(OTUdf4),
                               colData = metadata2,
                               design = ~Group*Timepoint)

SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than 4 for us to use vst
print(sizeFactors(SF.gdds)) #View size factors

#Our size factors are all less than 4, so we can use VST! VST = variance stabilizing transformation to minimize effects of small counts and normalize wrt library size

gvst <- varianceStabilizingTransformation(gdds, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst), 3) #view transformed gene count data

#generate as matrix 
matrix<-t(assay(gvst))
matrix<-as.data.frame(matrix)

matrix2 <- tibble::rownames_to_column(matrix, "Sample")
matrix3 <- merge(matrix2, metadata3, by = "Sample") %>% dplyr::select(-Sample)
rownames(matrix3) <- paste0(matrix3$ColDay)
OTUdf_final <- matrix3 %>% dplyr::select(-ColDay)

# Export as CSV for input into DIABLO
write.csv(OTUdf_final, "output/16S/processed_data/ASV_df.csv")

### Getting Taxonomy information for ASV Assignment 

tax.clean3 <- tibble::rownames_to_column(tax.clean2, "ASV_ID")

ASV_TAX <- merge(ps.rarefied_names2, tax.clean3, by = "ASV_ID")
write.csv(ASV_TAX, "output/16S/processed_data/ASV_taxID.csv")

```



# Alpha Diversity

Alpha diversity metrics assess the species diversity within the ecosystems, telling you how diverse a sequenced community is.

## Calulcating and plotting all Alpha diversity metrics

```{r, echo=FALSE, warning = FALSE}

plot_richness(ps.rarefied, x="GroupDay", color="GroupDay", measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher")) +
  geom_boxplot() +
  theme_classic() +  
  geom_point(aes(fill=GroupDay, group = GroupDay), pch = 21, size=2, position = position_jitterdodge(0.1)) +
  theme(axis.title.y = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0))) +
  theme(axis.title.x = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0))) +
  theme(axis.text.x.bottom = element_text(angle = -45)) 
```

## Calulating statistics for Shannon

```{r, echo=FALSE, warning = FALSE}
# wilcox.shannon <- pairwise.wilcox.test(rich$Shannon, 
#                                        sample_data(ps.rarefied)$GroupDay, 
#                                        p.adjust.method = "BH")
# tab.shannon <- wilcox.shannon$p.value %>%
#   as.data.frame() %>%
#   tibble::rownames_to_column(var = "group1") %>%
#   gather(key="group2", value="p.adj", -group1) %>%
#   na.omit()
# tab.shannon

# estimate richness function for all parameters shown above 
rich = estimate_richness(ps.rarefied, measures = c("Observed", "Shannon"))

# linear mixed effect model approach
Shannon_lmer <- aov(rich$Shannon ~ sample_data(ps.rarefied)$Group * sample_data(ps.rarefied)$Day)
Shannon_summary<- tidy(Shannon_lmer)
Shannon_summary

TukeyHSD(Shannon_lmer)

```

## Plotting Shannon

```{r}

p <- plot_richness(ps.rarefied, x="Day", color="Group", measures="Shannon") +
  geom_boxplot(aes(fill = Group), color = "black", width=.5, outlier.shape= NA, position = position_dodge(width = 0.5), alpha = 0.7) +
  stat_summary(fun=mean, geom="line", aes(group=Group, color = Group), position = position_dodge(width = 0.5))  + 
  geom_point(aes(fill = Group), color = "black", pch = 21, position=position_jitterdodge(dodge.width=0.5)) +
  scale_fill_manual(values=c("#CC3300", "#46008B", "#468B00")) +
  scale_color_manual(values=c("#CC3300", "#46008B", "#468B00")) + 
  # geom_text(x=52, y=2, label=paste("p(Day)=", Shannon_summary$`p.value`[1]), size=7, color=ifelse(Shannon_summary$`p.value`[1] < 0.05, "black", "darkgray")) + 
  # geom_text(x=0.65, y=-0.425, label=paste("p(Group)=", Shannon_summary$`p.value`[2]), size=7, color=ifelse(Shannon_summary$`p.value`[2] < 0.05, "black", "darkgray")) + 
  # geom_text(x=0.65, y=-0.45, label=paste("p(Day x Group)=", Shannon_summary$`p.value`[3]), size=7, color=ifelse(Shannon_summary$`p.value`[3] < 0.05, "black", "darkgray")) + 
  xlab("Timepoint") +  ylab("Shannon Diversity") + #Axis titles
  theme_bw() + 
    theme(legend.text = element_text(size=8), 
        legend.position="none",
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.margin = margin(1, 1, 1, 1, "cm"),
        axis.text = element_text(size=18), 
        panel.border = element_rect(color="black", fill=NA, size=0.75),
        title = element_text(size=25, face="bold"), 
        axis.title = element_text(size=18))

p$layers <- p$layers[-1]
p

ggsave(filename="output/16S/Shannon_div.jpeg", plot=p, dpi=300, width=12, height=12, units="in")

```

# PCoA, NMDS, and PERMANOVA/ADONIS

## PCoA 
```{r}

# creating a distance matrix with bray curtis method 
dist = phyloseq::distance(ps.rarefied, method="bray")

# creating an ordination from the same matrix 
ordination_PCOA = ordinate(ps.rarefied, method="PCoA", distance=dist)
ordination_NMDS = ordinate(ps.rarefied, method="NMDS", distance=dist)

# Plotting
pcoa <- plot_ordination(ps.rarefied, ordination_PCOA, color="Group") + 
  geom_point(size = 4, alpha=0.2, aes(shape = Day))+
  scale_colour_manual(values=c("#46008B", "#8B0046", "#468B00")) +
  scale_fill_manual(values=c("#46008B", "#8B0046", "#468B00")) + 
  scale_shape_manual(values=c(15, 17, 19)) +
  theme_classic() +
  ggtitle("PCoA Weighted Unifrac") + 
  geom_polygon(aes(fill=Group, alpha = 0.2)) +
#  stat_ellipse(aes(color=Group, group=Day),linetype = 2) +
  theme(plot.title = element_text(face = "bold")) +
  theme(strip.background = element_blank())
```


## PERMANOVA
```{r, echo=FALSE, warning = FALSE}
metadata <- data.frame(sample_data(ps.rarefied))
test.adonis <- adonis(dist ~ Group*Day, data = metadata)
test.adonis <- as.data.frame(test.adonis$aov.tab)
test.adonis
```

## NMDS
```{r}
## Finding hulls
library(plyr)
library(tibble)
df <- tibble::rownames_to_column(df, "VALUE")
df <- as.data.frame(as.matrix(ordination_NMDS$points))

df <- tibble::rownames_to_column(df, "Sample_ID")
SAMPLE_2 <- as.data.frame(as.matrix(SAMPLE))
SAMPLE_2 <- tibble::rownames_to_column(SAMPLE_2, "Sample_ID")

df2 <- merge(df, SAMPLE_2, by = "Sample_ID") 

find_hull <- function(df2) df2[chull(df2$MDS1, df2$MDS2), ]
hulls <- ddply(df2, "GroupDay", find_hull)

hulls <- dplyr::rename(hulls, NMDS1 = MDS1, NMDS2 = MDS2)

## Finding centroids
nmds.centroids<- df2%>% 
  dplyr::select(Day, Group, MDS1, MDS2)%>%
  dplyr::group_by(Day, Group)%>%
  dplyr::summarise(MDS1.mean = mean(MDS1),
                   MDS2.mean = mean(MDS2))

## Plotting NMDS

nmds <- plot_ordination(ps.rarefied, ordination_NMDS, color="Group", shape = "Day") + 
  geom_point(size = 4, alpha = 0.2, aes(shape = Day))+
  scale_colour_manual(values=c("#8B0046", "#46008B", "#468B00")) +
  scale_fill_manual(values=c("#8B0046","#46008B", "#468B00")) + 
  scale_shape_manual(values=c(15, 17, 19)) +
  theme_classic() +
#  geom_polygon(aes(fill=Group, alpha = 0.2)) +
#  stat_ellipse(type = "norm") +
    geom_text(x=0.65, y=-0.4, label=paste("p(Day)=", test.adonis$`Pr(>F)`[1]), size=7, color=ifelse(test.adonis$`Pr(>F)`[1] < 0.05, "black", "darkgray")) + 
  geom_text(x=0.65, y=-0.425, label=paste("p(Group)=", test.adonis$`Pr(>F)`[2]), size=7, color=ifelse(test.adonis$`Pr(>F)`[2] < 0.05, "black", "darkgray")) + 
  geom_text(x=0.65, y=-0.45, label=paste("p(Day x Group)=", test.adonis$`Pr(>F)`[3]), size=7, color=ifelse(test.adonis$`Pr(>F)`[3] < 0.05, "black", "darkgray")) + 
  annotate(geom="text", x=-0.1, y=-0.45, label="Stress = 0.106", color="black", size = 8) +
  theme(legend.text = element_text(size=8), 
        legend.position="none",
        plot.background = element_blank(),
        legend.title = element_text(size=10), 
        plot.margin = margin(1, 1, 1, 1, "cm"),
        axis.text = element_text(size=18), 
        panel.border = element_rect(color="black", fill=NA, size=0.75),
        title = element_text(size=25, face="bold"), 
        axis.title = element_text(size=18))

# Add centroids

nmdscen <- nmds + geom_polygon(data = hulls, alpha = 0.2, aes(color = Group, fill = Group, lty = Day)) +
  geom_point(aes(x=MDS1.mean, y=MDS2.mean,color=Group, shape = Day), data=nmds.centroids, size=4, show.legend=FALSE) + 
  scale_linetype_manual(values = c("solid", "dashed", "dotted")) +
  scale_colour_manual(values=c("#46008B", "#8B0046", "#468B00"), breaks = c("Control","Bleached", "Mortality"), labels = c("Control", "Bleached", "Partial Mortality")) +
  scale_fill_manual(values=c("#46008B", "#8B0046", "#468B00"), breaks = c("Control","Bleached", "Mortality"), labels = c("Control", "Bleached", "Partial Mortality")) + 
  scale_shape_manual(values=c(15, 17, 19)) +
  theme(legend.text = element_text(size=8), 
        legend.position="none",
        plot.background = element_blank(),
        legend.title = element_text(size=10), 
        plot.margin = margin(1, 1, 1, 1, "cm"),
        axis.text = element_text(size=18), 
        title = element_text(size=25, face="bold"), 
        axis.title = element_text(size=18))

# add segments

nmds.centroids$Day <- as.factor(nmds.centroids$Day)
nmds.centroids$Group <- as.factor(nmds.centroids$Group)

segpoints<-nmds.centroids%>%
  gather(variable, value, -(Day:Group)) %>%
  unite(temp, Day, variable) %>%
  spread(temp, value) 
  
names(segpoints)[2] <- "Day0_MDS1.mean"
names(segpoints)[3] <- "Day0_MDS2.mean"
names(segpoints)[4] <- "Day37_MDS1.mean"
names(segpoints)[5] <- "Day37_MDS2.mean"
names(segpoints)[6] <- "Day52_MDS1.mean"
names(segpoints)[7] <- "Day52_MDS2.mean"

nmdsfull<- nmdscen + 
  geom_segment(aes(x = Day0_MDS1.mean, y = Day0_MDS2.mean, xend = Day37_MDS1.mean, yend = Day37_MDS2.mean, colour = Group), data = segpoints, size=2, show.legend=FALSE, inherit.aes = FALSE) +
  geom_segment(aes(x = Day37_MDS1.mean, y = Day37_MDS2.mean, xend = Day52_MDS1.mean, yend = Day52_MDS2.mean, colour = Group), data = segpoints, size=2, arrow = arrow(length=unit(0.5,"cm")), show.legend=FALSE, inherit.aes = FALSE)

nmdsfull

ggsave(filename="output/16S/NMDS.jpeg", plot=nmdsfull, dpi=300, width=12, height=12, units="in")

```


## Pairwise PERMANOVA
- test using jaccard method later
```{r, echo=FALSE, warning = FALSE}
cbn <- combn(x=unique(metadata$GroupDay), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
ps.subs <- subset_samples(ps.rarefied, GroupDay %in% cbn[,i])
metadata_sub <- data.frame(sample_data(ps.subs))
permanova_pairwise <- adonis(phyloseq::distance(ps.subs, method = "bray") ~ GroupDay, 
                             data = metadata_sub)
p <- c(p, permanova_pairwise$aov.tab$`Pr(>F)`[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)
p.table
```

## Beta dispersion

```{r}
# input is the distance matrix calculated above 
beta.disperson <- vegan::betadisper(dist, phyloseq::sample_data(ps.rarefied)$GroupDay)

plot(beta.disperson, main = "Ordination Centroids and Dispersion", sub = "")

boxplot(beta.disperson)

# permutation test
permutest(beta.disperson)
```


# Abundance Barplots

Phylum
```{r, echo=FALSE, warning = FALSE}

ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Phylum', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Phylum <- as.character(ps.melt$Phylum)

ps.melt <- ps.melt %>%
  dplyr::group_by(Group, Day, Phylum) %>%
  dplyr::mutate(median=median(Abundance))
# select group median > 1
keep <- unique(ps.melt$Phylum[ps.melt$median > 1])
ps.melt$Phylum[!(ps.melt$Phylum %in% keep)] <- "< 1%"
#to get the same rows together
ps.melt_sum_phylum <- ps.melt %>%
  dplyr::group_by(Sample,Group, Day, Phylum) %>%
  dplyr::summarise(Abundance=sum(Abundance))

ggplot(ps.melt_sum_phylum, aes(x = Sample, y = Abundance, fill = Phylum)) + 
  geom_bar(stat = "identity", aes(fill=Phylum)) + 
  labs(x="", y="%") +
  facet_wrap(~Group+Day, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))

# Heatmap 
ggplot(ps.melt_sum_phylum, aes(x = Sample, y = Phylum)) + geom_tile(aes(fill = Abundance)) +
  scale_fill_distiller("Abundance", palette = "Blues", direction=1) + theme_classic() + 
  theme(axis.title.x = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0), face='bold', size = 12)) +
  theme(axis.title.y = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0), face='bold', size = 12)) + 
  theme(axis.text.x.bottom = element_text(angle = -90)) +
  theme(axis.text.y = element_text(colour = 'black', size = 10, face = 'italic')) + 
  theme(strip.text.x = element_text(size = 14, face='bold')) +
  facet_grid(~Day+Group, scales = "free") 

```


Family
```{r, echo=FALSE, warning = FALSE}
ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Family', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Family <- as.character(ps.melt$Family)

ps.melt <- ps.melt %>%
  dplyr::group_by(Group, Day, Family) %>%
  dplyr::mutate(median=median(Abundance))
# select group mean > 1
keep <- unique(ps.melt$Family[ps.melt$median > 2.5])
ps.melt$Family[!(ps.melt$Family %in% keep)] <- "< 2.5%"
#to get the same rows together
ps.melt_sum_family <- ps.melt %>%
  dplyr::group_by(Colony_ID, Group, Day, Family) %>%
  dplyr::summarise(Abundance=sum(Abundance))

ggplot(ps.melt_sum_family, aes(x = Colony_ID, y = Abundance, fill = Family)) + 
  geom_bar(stat = "identity", aes(fill=Family)) + 
  labs(x="", y="%") +
  facet_wrap(~Group+Day, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(legend.position = "right", 
        strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))

# Heatmap 
rel_fam<- ggplot(ps.melt_sum_family, aes(x = Colony_ID, y = Family)) + geom_tile(aes(fill = Abundance)) +
  scale_fill_distiller("Abundance", palette = "Blues", direction=1) + theme_classic() + 
  theme(axis.title.x = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0), face='bold', size = 12)) +
  theme(axis.title.y = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0), face='bold', size = 12)) + 
  theme(axis.text.x.bottom = element_text(angle = -45)) +
  theme(axis.text.y = element_text(colour = 'black', size = 14, face = 'italic')) + 
  theme(strip.text.x = element_text(size = 14, face='bold')) +
  facet_grid(~Group+Day, scales = "free") 

ggsave(filename="output/16S/Rel_Abun_Family.png", plot=rel_fam, dpi=300, width=12, height=8, units="in")


#Boxplots

ps.melt_sum_family$Family_F = factor(ps.melt_sum_family$Family, levels=c('Rhodobacteraceae','Paraspirulinaceae','Kiloniellaceae','Flavobacteriaceae', 'Endozoicomonadaceae', 'Cyclobacteriaceae', 'Alteromonadaceae', '< 2.5%'))

rel_fam_box <- ggplot(ps.melt_sum_family, aes(x=Day, y=Abundance, fill = Group)) +
  geom_boxplot(width=.5, outlier.shape= NA, position = position_dodge(width = 0.5), alpha = 0.7) +
  stat_summary(fun=mean, geom="line", aes(group=Group, color = Group), position = position_dodge(width = 0.5))  + 
  #  stat_summary(fun=mean, geom="point")
  geom_point(pch = 21, position=position_jitterdodge(dodge.width=0.5)) +
#    ylim(0,1) +
  scale_fill_manual(values=c("#CC3300", "#46008B", "#468B00")) +
  scale_color_manual(values=c("#CC3300", "#46008B", "#468B00")) + 
  scale_x_discrete(labels=c("0" = "Day 0", "37" = "Day 37", "52" = "Day 52")) +
  scale_y_continuous(position = "right") + 
  xlab("Timepoint") +  ylab("Percent Relative Abundance") + #Axis titles
  theme_bw() + theme(panel.border = element_rect(color="black", fill=NA, size=0.75), panel.grid.major = element_blank(), #Makes background theme white
                     panel.grid.minor = element_blank(), axis.line = element_blank()) +
  theme(axis.text = element_text(size = 15, color = "black"),
        axis.title = element_text(size = 18, color = "black", face="bold")) +
  facet_grid(Family_F ~ ., scales = "free", switch = "y") +
  theme(strip.text.y = element_text(size = 13, face="bold"), legend.position = "none")

rel_fam_box

ggsave(filename="output/16S/Rel_Abun_Family_Box.png", plot=rel_fam_box, dpi=300, width=7, height=15, units="in")


# ANOVA for each Family group

Rhodobacteraceae_data<- ps.melt_sum_family %>% filter(Family_F == "Rhodobacteraceae") 
Rhodobacteraceae_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = Rhodobacteraceae_data)
anova(Rhodobacteraceae_lmer, type="III")
capture.output(anova(Rhodobacteraceae_lmer, type="III"), file = "output/Statistics/Rhodobacteraceae_lmer.csv")

Paraspirulinaceae_data<- ps.melt_sum_family %>% filter(Family_F == "Paraspirulinaceae") 
Paraspirulinaceae_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = Paraspirulinaceae_data)
anova(Paraspirulinaceae_lmer, type="III")
capture.output(anova(Paraspirulinaceae_lmer, type="III"), file = "output/Statistics/Paraspirulinaceae_lmer.csv")

Kiloniellaceae_data<- ps.melt_sum_family %>% filter(Family_F == "Kiloniellaceae") 
Kiloniellaceae_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = Kiloniellaceae_data)
anova(Kiloniellaceae_lmer, type="III")
capture.output(anova(Kiloniellaceae_lmer, type="III"), file = "output/Statistics/Kiloniellaceae_lmer.csv")

Flavobacteriaceae_data<- ps.melt_sum_family %>% filter(Family_F == "Flavobacteriaceae") 
Flavobacteriaceae_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = Flavobacteriaceae_data)
anova(Flavobacteriaceae_lmer, type="III")
capture.output(anova(Flavobacteriaceae_lmer, type="III"), file = "output/Statistics/Flavobacteriaceae_lmer.csv")

Endozoicomonadaceae_data<- ps.melt_sum_family %>% filter(Family_F == "Endozoicomonadaceae") 
Endozoicomonadaceae_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = Endozoicomonadaceae_data)
anova(Endozoicomonadaceae_lmer, type="III")
capture.output(anova(Endozoicomonadaceae_lmer, type="III"), file = "output/Statistics/Endozoicomonadaceae_lmer.csv")

Cyclobacteriaceae_data<- ps.melt_sum_family %>% filter(Family_F == "Cyclobacteriaceae") 
Cyclobacteriaceae_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = Cyclobacteriaceae_data)
anova(Cyclobacteriaceae_lmer, type="III")
capture.output(anova(Cyclobacteriaceae_lmer, type="III"), file = "output/Statistics/Cyclobacteriaceae_lmer.csv")

L2.5_data<- ps.melt_sum_family %>% filter(Family_F == "< 2.5%") 
L2.5_lmer <- lmer(Abundance~Group*Day+(1|Colony_ID), data = L2.5_data)
L2.5_aov <- anova(L2.5_lmer, type="III")
capture.output(anova(L2.5_lmer, type="III"), file = "output/Statistics/L2.5_lmer.csv")

```


Genus
```{r, echo=FALSE, warning = FALSE}
ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Genus', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Genus <- as.character(ps.melt$Genus)

ps.melt <- ps.melt %>%
  group_by(Group, Day, Genus) %>%
  mutate(median=median(Abundance))
# select group mean > 1
keep <- unique(ps.melt$Genus[ps.melt$median > 2.5])
ps.melt$Genus[!(ps.melt$Genus %in% keep)] <- "< 2.5%"
#to get the same rows together
ps.melt_sum_genus <- ps.melt %>%
  group_by(Colony_ID, Group, Day, Genus) %>%
  dplyr::summarise(Abundance=sum(Abundance))

ggplot(ps.melt_sum_genus, aes(x = Colony_ID, y = Abundance, fill = Genus)) + 
  geom_bar(stat = "identity", aes(fill=Genus)) + 
  labs(x="", y="%") +
  facet_wrap(~Group+Day, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(legend.position = "right", 
        strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))

# Heatmap 
ggplot(ps.melt_sum_genus, aes(x = Colony_ID, y = Genus)) + geom_tile(aes(fill = Abundance)) +
  scale_fill_distiller("Abundance", palette = "Blues", direction=1) + theme_classic() + 
  theme(axis.title.x = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0), face='bold', size = 12)) +
  theme(axis.title.y = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0), face='bold', size = 12)) + 
  theme(axis.text.x.bottom = element_text(angle = -90)) +
  theme(axis.text.y = element_text(colour = 'black', size = 10, face = 'italic')) + 
  theme(strip.text.x = element_text(size = 14, face='bold')) +
  facet_grid(~Group+Day, scales = "free") 

```
