---
title: "16_Analysis_QIMME2"
author: "Kevin Wong"
date: "24/02/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Following this tutorial: https://www.yanh.org/2021/01/01/microbiome-r/

# Installing Packages

```{r}
p1 <- c("tidyverse", "vegan", "BiocManager")
p2 <- c("phyloseq", "ANCOMBC", "DESeq2", "ComplexHeatmap")
load_package <- function(p) {
  if (!requireNamespace(p, quietly = TRUE)) {
    ifelse(p %in% p1, 
           install.packages(p, repos = "http://cran.us.r-project.org/"), 
           BiocManager::install(p))
  }
  library(p, character.only = TRUE, quietly = TRUE)
}
invisible(lapply(c(p1,p2), load_package))

```

# Build phyloseq project

```{r, echo=FALSE, warning = FALSE}

otu <- read.table(file = "output/16S/processed_data/feature-table.tsv", sep = "\t", header = T, row.names = 1, skip = 1, comment.char = "")

taxonomy <- read.table(file = "output/16S/processed_data/taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

# clean the taxonomy, Greengenes format
tax <- taxonomy %>%
  dplyr::select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}

tax.clean2 <- tax.clean %>% 
  filter(Kingdom == "Bacteria") %>% 
  filter(Order != "Chloroplast")

metadata <- read.table(file = "output/16S/metadata.tsv", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)
TAX = tax_table(as.matrix(tax.clean2))
SAMPLE <- sample_data(metadata)
SAMPLE$Timepoint <- as.factor(SAMPLE$Timepoint)
SAMPLE$Group <- as.factor(SAMPLE$Group)
names(SAMPLE)[3] <- "Day"
SAMPLE$GroupDay <- paste(SAMPLE$Group, SAMPLE$Day, sep = "-")
TREE = read_tree("output/16S/processed_data/tree.nwk")
# merge the data
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)
ps

```

# Check the sequencing depth with rarefaction curves

From the QIIME2 output, there are no samples that drop off with sampling depth. 2000 and is the sample size that keeps all but 1 of my samples after testing

```{r, echo=FALSE, warning = FALSE}
set.seed(111) # keep result reproductive
ps.rarefied = rarefy_even_depth(ps, rngseed=1, sample.size=2300, replace=F)
ps.rarefied
```

# Creating a dataframe for omic integration

Renaming ASV ID and sample names. There is probably a more elegant way to do this...

```{r, echo=FALSE, warning = FALSE}
#Assigning ASV numbers to ASV ID
ps.rarefied_names <- as.data.frame(taxa_names(ps.rarefied))
names(ps.rarefied_names)[1] <- "ASV_ID"
ps.rarefied_names$Rownum <- as.numeric(rownames(ps.rarefied_names))
ps.rarefied_names$ASV_ASV <- "ASV"
ps.rarefied_names$ASV <- paste(ps.rarefied_names$ASV_ASV, ps.rarefied_names$Rownum, sep = "-")
ps.rarefied_names2 <- ps.rarefied_names %>% dplyr::select(ASV_ID, ASV)

# Extract abundance matrix from the phyloseq object
OTU1 = as(otu_table(ps.rarefied), "matrix")
# transpose if necessary
#if(taxa_are_rows(ps.rarefied)){OTU1 <- t(OTU1)}
# Coerce to data.frame
OTUdf = as.data.frame(OTU1)
#Make rownames into a column
OTUdf2 <- tibble::rownames_to_column(OTUdf, "ASV_ID")

# Merging dataframes to get ASV
OTUdf3 <- merge(OTUdf2, ps.rarefied_names2, by = "ASV_ID") %>% dplyr::select(-ASV_ID)
rownames(OTUdf3) <- paste0(OTUdf3$ASV)
OTUdf4 <- OTUdf3 %>% dplyr::select(-ASV)

#Formatting the Sample Datasheet to rename the samples
metadata2 <- tibble::rownames_to_column(metadata, "Sample") 
metadata2$ColDay <- paste(metadata2$Colony_ID, metadata2$Timepoint, sep = "-")
metadata3 <- metadata2 %>% dplyr::select(Sample, ColDay)

#Reformatting OTU dataframe and merge with sample info
OTUdf5 <- as.data.frame(t(OTUdf4))
OTUdf6 <- tibble::rownames_to_column(OTUdf5, "Sample")
OTUdf7 <- merge(OTUdf6, metadata3, by = "Sample") %>% dplyr::select(-Sample)
rownames(OTUdf7) <- paste0(OTUdf7$ColDay)
OTUdf_final <- OTUdf7 %>% dplyr::select(-ColDay)

# Export as CSV for input into DIABLO
write.csv(OTUdf_final, "output/16S/processed_data/ASV_df.csv")

```



# Alpha Diversity

Alpha diversity metrics assess the species diversity within the ecosystems, telling you how diverse a sequenced community is.

```{r, echo=FALSE, warning = FALSE}
plot_richness(ps.rarefied, x="GroupDay", measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), nrow=7) +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```


```{r, echo=FALSE, warning = FALSE}
rich = estimate_richness(ps.rarefied, measures = c("Observed", "Shannon"))
wilcox.observed <- pairwise.wilcox.test(rich$Observed, 
                                        sample_data(ps.rarefied)$GroupDay, 
                                        p.adjust.method = "BH")
tab.observed <- wilcox.observed$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.observed
```

```{r, echo=FALSE, warning = FALSE}
wilcox.shannon <- pairwise.wilcox.test(rich$Shannon, 
                                       sample_data(ps.rarefied)$GroupDay, 
                                       p.adjust.method = "BH")
tab.shannon <- wilcox.shannon$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.shannon
```

# PCoA and PERMANOVA/ADONIS

PCoA
```{r, echo=FALSE, warning = FALSE}
dist = phyloseq::distance(ps.rarefied, method="bray")
ordination = ordinate(ps.rarefied, method="PCoA", distance=dist)
plot_ordination(ps.rarefied, ordination, color="Group", shape = "Day") + 
  theme_classic() +
  theme(strip.background = element_blank())
```


PERMANOVA
```{r, echo=FALSE, warning = FALSE}
metadata <- data.frame(sample_data(ps.rarefied))
test.adonis <- adonis(dist ~ Group*Day, data = metadata)
test.adonis <- as.data.frame(test.adonis$aov.tab)
test.adonis
```


Pairwise PERMANOVA
- test using jaccard method later
```{r, echo=FALSE, warning = FALSE}
cbn <- combn(x=unique(metadata$GroupDay), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
ps.subs <- subset_samples(ps.rarefied, GroupDay %in% cbn[,i])
metadata_sub <- data.frame(sample_data(ps.subs))
permanova_pairwise <- adonis(phyloseq::distance(ps.subs, method = "bray") ~ GroupDay, 
                             data = metadata_sub)
p <- c(p, permanova_pairwise$aov.tab$`Pr(>F)`[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)
p.table
```

# NMDS and ANOSIM

NMDS, rememeber to report stress value
```{r, echo=FALSE, warning = FALSE}
dist = phyloseq::distance(ps.rarefied, method="bray", binary = TRUE)

ordination = ordinate(ps.rarefied, method="NMDS", distance=dist)
plot_ordination(ps.rarefied, ordination, color="Group", shape = "Day") + 
  theme_classic() +
  theme(strip.background = element_blank())
```

NMDS modified
```{r, echo=FALSE, warning = FALSE}
p2 = plot_ordination(ps.rarefied, ordination, type="Colony_ID", color="Group", shape="Day") 
p2 + geom_polygon(aes(fill=Group)) + geom_point(size=5) + ggtitle("samples") 


```


ANOSIM
```{r, echo=FALSE, warning = FALSE}
metadata <- data.frame(sample_data(ps.rarefied))
anosim(dist, metadata$GroupDay)
```

Pairwise ANOSIM
```{r, echo=FALSE, warning = FALSE}

cbn <- combn(x=unique(metadata$GroupDay), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
ps.subs <- subset_samples(ps.rarefied, GroupDay %in% cbn[,i])
metadata_sub <- data.frame(sample_data(ps.subs))
permanova_pairwise <- anosim(phyloseq::distance(ps.subs, method="jaccard", binary = TRUE), 
                             metadata_sub$GroupDay)
p <- c(p, permanova_pairwise$signif[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)
p.table
```

# Abundance Barplot

Phylum
```{r, echo=FALSE, warning = FALSE}

ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Phylum', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Phylum <- as.character(ps.melt$Phylum)

ps.melt <- ps.melt %>%
  group_by(Group, Day, Phylum) %>%
  mutate(median=median(Abundance))
# select group median > 1
keep <- unique(ps.melt$Phylum[ps.melt$median > 1])
ps.melt$Phylum[!(ps.melt$Phylum %in% keep)] <- "< 1%"
#to get the same rows together
ps.melt_sum <- ps.melt %>%
  group_by(Sample,Group, Day, Phylum) %>%
  summarise(Abundance=sum(Abundance))

ggplot(ps.melt_sum, aes(x = Sample, y = Abundance, fill = Phylum)) + 
  geom_bar(stat = "identity", aes(fill=Phylum)) + 
  labs(x="", y="%") +
  facet_wrap(~Group+Day, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))

```

Family
```{r, echo=FALSE, warning = FALSE}
ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Family', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Genus <- as.character(ps.melt$Family)

ps.melt <- ps.melt %>%
  group_by(Group, Day, Family) %>%
  mutate(median=median(Abundance))
# select group mean > 1
keep <- unique(ps.melt$Family[ps.melt$median > 2.5])
ps.melt$Family[!(ps.melt$Family %in% keep)] <- "< 2.5%"
#to get the same rows together
ps.melt_sum <- ps.melt %>%
  group_by(Colony_ID, Group, Day, Family) %>%
  summarise(Abundance=sum(Abundance))

ggplot(ps.melt_sum, aes(x = Colony_ID, y = Abundance, fill = Family)) + 
  geom_bar(stat = "identity", aes(fill=Family)) + 
  labs(x="", y="%") +
  facet_wrap(~Group+Day, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(legend.position = "right", 
        strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))
```


Genus
```{r, echo=FALSE, warning = FALSE}
ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
# agglomerate taxa
glom <- tax_glom(ps.rel, taxrank = 'Genus', NArm = FALSE)
ps.melt <- psmelt(glom)
# change to character for easy-adjusted level
ps.melt$Genus <- as.character(ps.melt$Genus)

ps.melt <- ps.melt %>%
  group_by(Group, Day, Genus) %>%
  mutate(median=median(Abundance))
# select group mean > 1
keep <- unique(ps.melt$Genus[ps.melt$median > 2.5])
ps.melt$Genus[!(ps.melt$Genus %in% keep)] <- "< 2.5%"
#to get the same rows together
ps.melt_sum <- ps.melt %>%
  group_by(Sample, Group, Day, Genus) %>%
  summarise(Abundance=sum(Abundance))

ggplot(ps.melt_sum, aes(x = Sample, y = Abundance, fill = Genus)) + 
  geom_bar(stat = "identity", aes(fill=Genus)) + 
  labs(x="", y="%") +
  facet_wrap(~Group+Day, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(legend.position = "right", 
        strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))
```
